alibi_detect/utils/pytorch/kernels.py

```diff
@@ -20,21 +56,307 @@ def sigma_median(x: torch.Tensor, y: torch.Tensor, dist: torch.Tensor) -> torch.
+        elif (isinstance(other, BaseKernel) or
+              isinstance(other, ProductKernel) or
+              isinstance(other, torch.Tensor)):
+            sum_kernel = SumKernel()
+            sum_kernel.kernel_list.append(self)
+            sum_kernel.kernel_list.append(other)
+            return sum_kernel
+        else:
+            raise ValueError('Kernels can only added to another kernel or a constant.')
+
+    def __radd__(self, other: Union['BaseKernel', 'SumKernel', 'ProductKernel']) -> 'SumKernel':
+        return self.__add__(other)
+
+    def __mul__(
+        self,
+        other: Union['BaseKernel', 'SumKernel', 'ProductKernel', torch.Tensor]
+    ) -> Union['SumKernel', 'ProductKernel']:
+        if isinstance(other, ProductKernel):
+            other.kernel_factors.append(self)
+            return other
+        elif isinstance(other, SumKernel):
+            sum_kernel = SumKernel()
+            for k in other.kernel_list:
+                sum_kernel.kernel_list.append(self * k)
+            return sum_kernel
+        else:
+            prod_kernel = ProductKernel()
+            prod_kernel.kernel_factors.append(self)
+            prod_kernel.kernel_factors.append(other)
+            return prod_kernel
+
+    def __rmul__(
+        self,
+        other: Union['BaseKernel', 'SumKernel', 'ProductKernel']
+    ) -> Union['SumKernel', 'ProductKernel']:
+        return self.__mul__(other)
+
+    def __truediv__(self, other: torch.Tensor) -> Union['SumKernel', 'ProductKernel']:
+        if isinstance(other, torch.Tensor):
+            return self.__mul__(1. / other)
+        else:
+            raise ValueError('Kernels can only be divided by a constant.')
+
+    def __rtruediv__(self, other):
+        raise ValueError('Kernels can not be used as divisor.')
+
+    def __sub__(self, other):
+        raise ValueError('Kernels do not support substraction.')
+
+    def __rsub__(self, other):
+        raise ValueError('Kernels do not support substraction.')
+
+
+class SumKernel(torch.nn.Module):
+    def __init__(self) -> None:
+        """
+        Construct a kernel by summing different kernels.
+        """
+        super().__init__()
+        self.kernel_list: List[Union[BaseKernel, SumKernel, ProductKernel, torch.Tensor]] = []
+
+    def forward(self, x: Union[np.ndarray, torch.Tensor], y: Union[np.ndarray, torch.Tensor],
+                infer_parameter: bool = False) -> torch.Tensor:
+        value_list: List[torch.Tensor] = []
+        for k in self.kernel_list:
+            if isinstance(k, BaseKernel) or isinstance(k, SumKernel) or isinstance(k, ProductKernel):
+                value_list.append(k(x, y, infer_parameter))
+            elif isinstance(k, torch.Tensor):
+                value_list.append(k * torch.ones((x.shape[0], y.shape[0])))
+            else:
+                raise ValueError(type(k) + 'is not supported by SumKernel.')
+        return torch.sum(torch.stack(value_list), dim=0)
+
+    def __add__(
```

Comment:
From here onwards we need lots of comments/docstrings for maintainability. Otherwise this might turn out to be a mess quite quickly.

Reply:
A key part would also be to clarify how `[m,n]` kernel matrices are spliced for mixed-type data with `active_dims` on subsets of the features (e.g. 3 numerical and 4 categorical) and where we use combined kernels (e.g. a product kernel).